### ISSUE TITLE
nth item

### ISSUE THREAD

**Issue by pylang** [erikrose/more-itertools#139]
I came across [this recipe](http://stackoverflow.com/a/8337489/4531270) by R. Hettinger:

```python
>>> x = [False,True,True,False,True,False,True,False,False,False,True,False,True]
>>> nth_item(50, True, x)
-1
>>> nth_item(0, True, x)
1
>>> nth_item(1, True, x)
2
>>> nth_item(2, True, x)
4
>>> nth_item(3, True, x)
6
```

Code

```python
>>> from itertools import compress, count, imap, islice
>>> from functools import partial
>>> from operator import eq

>>> def nth_item(n, item, iterable):
        indicies = compress(count(), imap(partial(eq, item), iterable))
        return next(islice(indicies, n, None), -1)
```

I thought this may be a useful addition.

**Comment by bbayles** [comment]
I was a bit confused by the name. We have [nth](https://more-itertools.readthedocs.io/en/latest/api.html#more_itertools.nth), which does return the nth item. However, this is "what is the index of the nth occurrence of a particular item?"

So with a better name I'm open to this. `index_of_nth_occurrence` isn't very snappy?

**Comment by pylang** [comment]
I agree.  `nth_index` makes more sense.

**Comment by pylang** [comment]
`locate(n, item, iterable)` , as in locate the `n`th `item` in this `iterable`?  Or `where`?

Numpy has a [`where`][2] method I believe.  They also have [`argwhere`][3] which more similar to this application.  Not sure if it is general enough to call `index`... I'm not attached to any of these names.  

   [2]: https://docs.scipy.org/doc/numpy-1.12.0/reference/generated/numpy.where.html
   [3]: https://docs.scipy.org/doc/numpy/reference/generated/numpy.argwhere.html

**Comment by bbayles** [comment]
I've asked Raymond Hettinger on Twitter if he would bless the inclusion of this.

The implementation is dynamite, but I think we'll want to get the name and docstring right. Making "nth occurrence" be interpreted as 0-based is confusing in a way that "nth item" is not. A co-worker of mine suggested "find the index of the item, skipping the first n matches."

I might also suggest changing this to: `locate(iterable, pred, n=0)` to be more flexible. That is, "find the first index where `pred` is `True`, skipping the first `n` matches. 

Thinking more generally, I wouldn't mind writing some more itertools that deal with locating indexes. Consider:
* `index_subsequence(iterable, sub, start=0, end=None)` - find the index where the subsequence `sub` occurs in `iterable`. Like `str.index()`, but for generic iterables and no negative indexes.
* `index_item(iterable, item, start=0, end=None)` - find the index where the item `item` occurs in `iterable`. Like `list.index()`, but for generic iterables and no negative indexes.

**Comment by pylang** [comment]
**Re: blessings**
Good call on pinging R. Hettinger.  That seems appropriate.

**Re: names and docstring**
You spotted two things that I would also change, 0-indexed `n` and the signature.  

 Looking at existing tools, there seems to be some variance in how "nth" is interpreted in the recipes.  The  [`nth`][1] tool feels 0-indexed:
```python
Returns the nth item or a default value

>>> l = range(10)
>>> nth(l, 3)
3

```

Although other recipes ([`take`][2] and [`tail`][3]) seem to "go up to `n`", which happens to equal the length:

```python
Return first n items of the iterable as a list

>>> take(3, range(10))
[0, 1, 2]
```

Since the idea of a "zeroth occurrence" of `True` is lacks meaning, I'm in favor of the second interpretation.  +1 for the new signature proposal.

**Re: more itertools** 

While I like the idea of having tools for handling indices (a common problem), I would prefer to minimize the number to tools.  I'm cautious of a possible categorical explosion of index-based tools if not sufficiently generalized, although you and other contributors may have better approaches for handling this.  If it turns out there are many cases for indexing, I wouldn't mind a class with methods (similar to `peek`), perhaps  `index()` item (by default) and then `index.subsequence()`, `index.nested()`, etc.

[1]: https://github.com/erikrose/more-itertools/blob/master/more_itertools/recipes.py#L145
[2]: https://github.com/erikrose/more-itertools/blob/master/more_itertools/recipes.py#L60
[3]: https://github.com/erikrose/more-itertools/blob/master/more_itertools/recipes.py#L93

**Comment by bbayles** [comment]
See [this branch](https://github.com/erikrose/more-itertools/compare/indexes?expand=1) for what I'm thinking of adding.

**Comment by pylang** [comment]
I'm still unsettled about adding extra index-based tools.  Thinking more generally, perhaps we restrict ourselves by looking for a particular index.  We could generalize further by yielding all indices that meet a condition.  I attempted to abstract `locate` and apply examples from PR#140.

```python
import itertools as it


def locate(iterable, pred=bool):
    """Yield all indicies for which the callable *pred* is ``True``."""
    yield from it.compress(it.count(), map(pred, iterable))

iterable = [0, 1, 1, 0, 1, 0, 0]
idxs = locate(iterable)

# Similiar to `nth_item` and present `locate`
next(idxs)
# Out: 1

# # Similar to `item_index`, skipped the first match
next(idxs)
# Out: 2

# Unique, generalized to find all indicies
idxs = locate(iterable)
list(idxs)
# Out: [1, 5]

# Similar to `item_index(iterable, item, start=3, end=6)`
iterable = ['a', 'b', 'c', 'd', 'a', 'b', 'c', 'd']
item = 'b'
[i for i in locate(iterable, pred=lambda x: x==item) if i > 3]
# Out: [5]

```

Since we can access all indices, we need not worry about skipping `n` or `start` and `end` keywords, but rather leave it to the user.   It's also nice that the implementation is already baked in Hettinger's recipe.  While similar to a list comprehension, I suspect a C-level speed up from the functional implementation (according to the [SO post][1]) plus the benefits of using generators.  `sub_index` still has utility, although I sense it is a coupling between `windowed` and generalized `locate`.  I'm curious on thoughts on this abstraction first. 

In short, `locate` all indices may be a better approach than get the `nth_item`.

  [1]: https://stackoverflow.com/questions/8337069/find-the-index-of-the-nth-item-in-a-list/8337489#8337489

**Comment by bbayles** [comment]
I like yielding the matching indexes in `locate`. Given that the other two apparently aren't faster than the naive loop versions they might not be worth including.

**Comment by bbayles** [comment]
Closed by PR #140; the new `locate` tool will go in the next release.
